src/Diagnostics.HealthChecks/CertificatesHealthCheck.cs

```diff
@@ -0,0 +1,131 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Diagnostics.HealthChecks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Omex.Extensions.Abstractions;
+using Microsoft.Omex.Extensions.Diagnostics.HealthChecks.Composables;
+using Microsoft.Omex.Extensions.Hosting.Certificates;
+
+namespace Microsoft.Omex.Extensions.Diagnostics.HealthChecks;
+
+/// <summary>
+/// This health check verifies whether certificates have been loaded correctly by checking their validity period and ensuring that they include a private key.
+/// </summary>
+public class CertificatesHealthCheck : IHealthCheck
+{
+	private readonly IHealthCheck m_healthCheck;
+
+	/// <summary>
+	/// Initializes a new instance of the <see cref="CertificatesHealthCheck"/> class.
+	/// </summary>
+	/// <param name="parameters">
+	/// The health check parameters passed at DI configuration time.
+	/// They can be used to specify escalation configuration (e.g. the IcM ticket severity, owning team),
+	/// </param>
+	/// <param name="logger">The logger.</param>
+	/// <param name="activitySource">The activity source.</param>
+	/// <param name="certificateReader">The certificate reader.</param>
+	/// <param name="options">Options</param>
+	public CertificatesHealthCheck(
+		HealthCheckParameters parameters,
+		ILogger<CertificatesHealthCheck> logger,
+		ActivitySource activitySource,
+		ICertificateReader certificateReader,
+		IOptions<CertificatesHealthCheckOptions> options)
+	{
+		HashSet<string> uniqueCertificateSubjectNames = new(options.Value.CertSubjectNames);
+
+		m_healthCheck = new CertificatesHealthCheckInternal(
+			parameters,
+			uniqueCertificateSubjectNames,
+			logger,
+			certificateReader)
+				.AsObservableHealthCheck(activitySource, logger, parameters: parameters);
```

Comment:
This health check should have only one responsibility which is ensuring certificates validity. Observable wrapper will be added by the consumer if desired.

Reply:
The AsObservableHealthCheck extension method only adds activity tracking and exception handling, which are not defined in the HealthCheck logic itself. If we remove the wrapper, the consumer side will be forced to add exact same part into their code, which could lead to unnecessary code duplication across projects. I suggest keeping it as it is to provide a ready-to-go solution.

Reply:
I have to agree regarding the registration pattern being wrong here. However, I don't think we should publicly expose the internal health check class, as that corrupts any effort to unify on the Observable health check pattern.
In my mind the CertificatesHealthCheckInternal class is the actual health check. A health check is a unit of logic that evaluates some aspect of system/service health.
I'd suggest following the extension-led pattern that is already defined for AddEndpointHttpHealthCheck, etc. (https://github.com/microsoft/Omex/blob/main/src/Diagnostics.HealthChecks.AspNetCore/HealthCheckComposablesExtensions.cs) where the health check is internal and the addition of the health check is done through publically exposed health check builder extensions.

Reply:
Http health check is something of a special case, it's important to set it up correctly, so we hid the complexity behind convenience methods. I don't really have a strong preference, one approach could be similar to what you suggest Keith, in line with multiple examples here: https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/tree/master/src.

Reply:
In my mind our preferred approach was <Custom health check in the service> -> <Observable> -> <Startup> -> <Library health check>. That would go against .AddXXX pattern.

Reply:
After chatting with Anna, these approaches are not exclusive, we can have a convenience .Add method for typical usage and also expose the class for custom handling on consumer side.

Reply:
I added .Add method and removed wrapping class